# Data Gathering

### programmatically downloaded using the Requests library

In [2]:
#import libraries
import requests
import pandas as pd
import json
import datetime as dt

In [4]:
#store file to file handler
r = requests.get('https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv')
code = r.status_code
print(code)
#save downloaded file to local folder
open('image-predictions.tsv', 'wb').write(r.content)



200


335079

### Download tweet ID, retweet count, and favorite count using Tweepy API

In [ ]:
#read twitte IDs

#read the twitter-archive-enhanced.csv file in to dataframe
df_1 = pd.read_csv('twitter-archive-enhanced.csv')
df_1
#convert the first column of the datafram(tweet id) to a list
IDlist = df_1['tweet_id'].tolist()
IDlist

In [ ]:
#pull twitter status json from twitter API and store them into a list
import tweepy

# key and token omitted for privacy
consumer_key = '*'
consumer_secret = '*'
access_token = '*'
access_secret = '*'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit  = True, wait_on_rate_limit_notify = True)

tweet_list = []

for tid in IDlist:
    try:
        tweet = api.get_status(tid, tweet_mode='extended')
        tweet_list.append(tweet._json)
    except:
        print('tweet ' + str(tid) + "doesn't" exist')



In [5]:
#check if all tweet status has been downloaded
len(tweet_list)

2348

In [ ]:
# write json to txt file

# type(tweet_json)
with open('tweet_json.txt', 'w') as outfile:
    for tweet_json in tweet_list:
        json.dump(tweet_json, outfile)
        outfile.write('\n') #add a newline character at the end of each json

In [5]:
# read json from text

tweet_df_raw = pd.read_json('tweet_json.txt', lines = 'True')
tweet_df = tweet_df_raw[['id', 'retweet_count', 'favorite_count']]

tweet_df

id  retweet_count  favorite_count
0     892420643555336193           8758           39245
1     892177421306343426           6422           33609
2     891815181378084864           4265           25317
3     891689557279858688           8849           42611
4     891327558926688256           9631           40774
5     891087950875897856           3206           20444
6     890971913173991426           2124           11987
7     890729181411237888          19374           66310
8     890609185150312448           4362           28058
9     890240255349198849           7604           32303
10    890006608113172480           7512           30964
11    889880896479866881           5077           28063
12    889665388333682689          10261           48185
13    889638837579907072           4659           27468
14    889531135344209921           2285           15265
15    889278841981685760           5575           25585
16    888917238123831296           4635           29391
17    888804989199671297           4484           25881
18    888554962724278272           3685           20155
19    888078434458587136           3607           22026
20    887705289381826560           5539           30535
21    887517139158093824          11949           46711
22    887473957103951883          18718           69931
23    887343217045368832          10629           34049
24    887101392804085760           6106           30887
25    886983233522544640           7967           35575
26    886736880519319552           3383           12203
27    886680336477933568           4571           22674
28    886366144734445568           3260           21371
29    886267009285017600              4             117
...                  ...            ...             ...
2318  666411507551481857            337             457
2319  666407126856765440             42             112
2320  666396247373291520             90             170
2321  666373753744588802             95             193
2322  666362758909284353            586             794
2323  666353288456101888             74             226
2324  666345417576210432            144             305
2325  666337882303524864             95             202
2326  666293911632134144            364             515
2327  666287406224695296             70             151
2328  666273097616637952             79             180
2329  666268910803644416             36             107
2330  666104133288665088           6778           14599
2331  666102155909144576             14              80
2332  666099513787052032             72             159
2333  666094000022159362             76             165
2334  666082916733198337             46             119
2335  666073100786774016            172             328
2336  666071193221509120             65             153
2337  666063827256086533            226             489
2338  666058600524156928             59             115
2339  666057090499244032            144             304
2340  666055525042405380            259             447
2341  666051853826850816            869            1242
2342  666050758794694657             59             135
2343  666049248165822465             40             110
2344  666044226329800704            143             305
2345  666033412701032449             46             126
2346  666029285002620928             47             131
2347  666020888022790149            523            2528

[2348 rows x 3 columns]

# Data Assessing

In [6]:
#load 3 dataframes from different sources
archive_df = pd.read_csv('twitter-archive-enhanced.csv')
image_df = pd.read_csv('image-predictions.tsv', sep='\t')
status_df = tweet_df.copy()
status_df.rename(columns = {"id": "tweet_id"}, inplace = True)

In [7]:
status_df

tweet_id  retweet_count  favorite_count
0     892420643555336193           8758           39245
1     892177421306343426           6422           33609
2     891815181378084864           4265           25317
3     891689557279858688           8849           42611
4     891327558926688256           9631           40774
5     891087950875897856           3206           20444
6     890971913173991426           2124           11987
7     890729181411237888          19374           66310
8     890609185150312448           4362           28058
9     890240255349198849           7604           32303
10    890006608113172480           7512           30964
11    889880896479866881           5077           28063
12    889665388333682689          10261           48185
13    889638837579907072           4659           27468
14    889531135344209921           2285           15265
15    889278841981685760           5575           25585
16    888917238123831296           4635           29391
17    888804989199671297           4484           25881
18    888554962724278272           3685           20155
19    888078434458587136           3607           22026
20    887705289381826560           5539           30535
21    887517139158093824          11949           46711
22    887473957103951883          18718           69931
23    887343217045368832          10629           34049
24    887101392804085760           6106           30887
25    886983233522544640           7967           35575
26    886736880519319552           3383           12203
27    886680336477933568           4571           22674
28    886366144734445568           3260           21371
29    886267009285017600              4             117
...                  ...            ...             ...
2318  666411507551481857            337             457
2319  666407126856765440             42             112
2320  666396247373291520             90             170
2321  666373753744588802             95             193
2322  666362758909284353            586             794
2323  666353288456101888             74             226
2324  666345417576210432            144             305
2325  666337882303524864             95             202
2326  666293911632134144            364             515
2327  666287406224695296             70             151
2328  666273097616637952             79             180
2329  666268910803644416             36             107
2330  666104133288665088           6778           14599
2331  666102155909144576             14              80
2332  666099513787052032             72             159
2333  666094000022159362             76             165
2334  666082916733198337             46             119
2335  666073100786774016            172             328
2336  666071193221509120             65             153
2337  666063827256086533            226             489
2338  666058600524156928             59             115
2339  666057090499244032            144             304
2340  666055525042405380            259             447
2341  666051853826850816            869            1242
2342  666050758794694657             59             135
2343  666049248165822465             40             110
2344  666044226329800704            143             305
2345  666033412701032449             46             126
2346  666029285002620928             47             131
2347  666020888022790149            523            2528

[2348 rows x 3 columns]

In [8]:
archive_df.info()
archive_df.sample()
type(archive_df.timestamp)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

pandas.core.series.Series

In [9]:
archive_df.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

issues in archive dataframe
1. Quality issues:
    - Retweets: some of the tweets in this dataframe are retweet, as mentioned in the project detail. These retweet are not supposed to be included in the analysis
    - Unnecessary information: text, sources are not needed for analysis. Retweeted_status_id and retweeted_status_user_id, and retweetd_status_timestamp are not needed after data cleaning procedure. 
    - The rating_numerator and rating_denominator can be combined into one column in decimal form.
    - Wrong data type for tweet id. Since no calculations will be applied on tweet ID, the tweet ID needs to be str instead of int64.
    - The timestamp column has wrong data type. 
2. Tidiess issues
    - Dog stages are not in one column, instead, they are divided into 4.
    - Date and time should be two variables for the purpose of analysis. 

In [10]:
image_df.info()
image_df.p1.value_counts()
#image_df.sample()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


golden_retriever             150
Labrador_retriever           100
Pembroke                      89
Chihuahua                     83
pug                           57
chow                          44
Samoyed                       43
toy_poodle                    39
Pomeranian                    38
malamute                      30
cocker_spaniel                30
French_bulldog                26
miniature_pinscher            23
Chesapeake_Bay_retriever      23
seat_belt                     22
German_shepherd               20
Siberian_husky                20
Staffordshire_bullterrier     20
Cardigan                      19
web_site                      19
beagle                        18
teddy                         18
Eskimo_dog                    18
Maltese_dog                   18
Shetland_sheepdog             18
Rottweiler                    17
Shih-Tzu                      17
Lakeland_terrier              17
kuvasz                        16
Italian_greyhound             16
          

issues in image dataframe
1. Quality issues:
    - Non-descriptive column headers: p1, p1_conf, p1_dog, p2, p2_dog, p3, p3_conf, p3_dog etc. 
    - Some of the dog breeds has first letter capitalized and some are not. 
    - Some of the dog breeds are misspelled (ie. 19 of the breeds is website) or intended misspelled (ie. cheeseburger) 
    
2. Tidiess issues

In [11]:
status_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2348 entries, 0 to 2347
Data columns (total 3 columns):
tweet_id          2348 non-null int64
retweet_count     2348 non-null int64
favorite_count    2348 non-null int64
dtypes: int64(3)
memory usage: 73.4 KB


issues in status dataframe
1. Quality issues:
    - wrong data type for tweet ID
2. Tidiess issues

### Summary
Quality issues:
    1. Retweets: some of the tweets in this dataframe are retweet. as mentioned in the project detail, These retweets are not supposed to be included in the analysis.
    2. Unnecessary information: text, sources are not needed for analysis. Retweeted_status_id and retweeted_status_user_id, and retweetd_status_timestamp are not needed after data cleaning procedure. 
    3. Wrong data type for tweet id. Since no calculations will be applied on tweet ID, the tweet ID needs to be str instead of int64.
    4. The timestamp column has wrong data type. 
    5. Non-descriptive column headers: p1, p1_conf, p1_dog, p2, p2_dog, p3, p3_conf, p3_dog etc. 
    6. Some of the dog breeds has first letter capitalized and some are not. 
    7. Many ratings' denominator are not 10, even the numerator which greater than 10 is the feature of this twitter account, keep the denominator same is vital to later analysis. 
    8. Missing data: the image prediction part dataframe has 2075 rows while the status df has 2348 and the archive has 2356. Observations that has missing image predictions are considered as missing data.

Tidyness issues:
    1. Dog stages are not in one column, instead, they are divided into 4.
    2. One observation unit does not form a table. At least, retweet_count and favorite_count needs to be part of the archive dataframe to form a complete observation unit. 
    *3. archive dataframe, time and dtae should be two variables instead of 1.

# Data Cleaning

Issue #1 : retweets

#### Define:
Remove all tweet rows that has nonnull value in retweeted_status_id column in archive dataframe.

#### Code

In [12]:
# create a copy
archive_df_clean = archive_df.copy()

# only keep tweet whose retweeted_status_id is NAn
archive_df_clean = archive_df_clean[archive_df_clean['retweeted_status_id'].isnull()]

#### Test

In [13]:
archive_df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2175 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2175 non-null object
source                        2175 non-null object
text                          2175 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2117 non-null object
rating_numerator              2175 non-null int64
rating_denominator            2175 non-null int64
name                          2175 non-null object
doggo                         2175 non-null object
floofer                       2175 non-null object
pupper                        2175 non-null object
puppo                         2175 non-null object
dtypes: float64(4), int64(3), object(1

Issue #2: Unnecessary information
#### Define:
Remove columns: expanded_urls, in_reply_to_status_id, in_reply_to_user_id, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, source. 

#### Code

In [14]:
# Drop unwanted columns
archive_df_clean = archive_df_clean.drop(columns=['expanded_urls', 'in_reply_to_status_id', 'in_reply_to_user_id', 'text', 'retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp', 'source'])

#### Test

In [15]:
list(archive_df_clean)

['tweet_id',
 'timestamp',
 'rating_numerator',
 'rating_denominator',
 'name',
 'doggo',
 'floofer',
 'pupper',
 'puppo']

Issue #3: Wrong datatype for tweet ID
#### Define:
Change data type of Tweet ID to string

#### Code

In [16]:
# change the datatype of tweet_id
archive_df_clean['tweet_id'] = archive_df_clean['tweet_id'].astype(str)
# change the datatype of tweet_id column in status_df as well
status_df['tweet_id'] = status_df['tweet_id'].astype(str)



#### Test

In [17]:
archive_df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 9 columns):
tweet_id              2175 non-null object
timestamp             2175 non-null object
rating_numerator      2175 non-null int64
rating_denominator    2175 non-null int64
name                  2175 non-null object
doggo                 2175 non-null object
floofer               2175 non-null object
pupper                2175 non-null object
puppo                 2175 non-null object
dtypes: int64(2), object(7)
memory usage: 169.9+ KB


Issue #4: Wrong datatype for time stamp column
#### Define:
Remove columns: change the timestamp column to 

#### Code

In [18]:
#analysis the timestamp column
type(archive_df_clean['timestamp'])

archive_df_clean.head(100)

tweet_id                  timestamp  rating_numerator  \
0    892420643555336193  2017-08-01 16:23:56 +0000                13   
1    892177421306343426  2017-08-01 00:17:27 +0000                13   
2    891815181378084864  2017-07-31 00:18:03 +0000                12   
3    891689557279858688  2017-07-30 15:58:51 +0000                13   
4    891327558926688256  2017-07-29 16:00:24 +0000                12   
5    891087950875897856  2017-07-29 00:08:17 +0000                13   
6    890971913173991426  2017-07-28 16:27:12 +0000                13   
7    890729181411237888  2017-07-28 00:22:40 +0000                13   
8    890609185150312448  2017-07-27 16:25:51 +0000                13   
9    890240255349198849  2017-07-26 15:59:51 +0000                14   
10   890006608113172480  2017-07-26 00:31:25 +0000                13   
11   889880896479866881  2017-07-25 16:11:53 +0000                13   
12   889665388333682689  2017-07-25 01:55:32 +0000                13   
13   889638837579907072  2017-07-25 00:10:02 +0000                12   
14   889531135344209921  2017-07-24 17:02:04 +0000                13   
15   889278841981685760  2017-07-24 00:19:32 +0000                13   
16   888917238123831296  2017-07-23 00:22:39 +0000                12   
17   888804989199671297  2017-07-22 16:56:37 +0000                13   
18   888554962724278272  2017-07-22 00:23:06 +0000                13   
20   888078434458587136  2017-07-20 16:49:33 +0000                12   
21   887705289381826560  2017-07-19 16:06:48 +0000                13   
22   887517139158093824  2017-07-19 03:39:09 +0000                14   
23   887473957103951883  2017-07-19 00:47:34 +0000                13   
24   887343217045368832  2017-07-18 16:08:03 +0000                13   
25   887101392804085760  2017-07-18 00:07:08 +0000                12   
26   886983233522544640  2017-07-17 16:17:36 +0000                13   
27   886736880519319552  2017-07-16 23:58:41 +0000                13   
28   886680336477933568  2017-07-16 20:14:00 +0000                13   
29   886366144734445568  2017-07-15 23:25:31 +0000                12   
30   886267009285017600  2017-07-15 16:51:35 +0000                12   
..                  ...                        ...               ...   
76   878057613040115712  2017-06-23 01:10:23 +0000                14   
77   877736472329191424  2017-06-22 03:54:17 +0000                13   
79   877556246731214848  2017-06-21 15:58:08 +0000                12   
80   877316821321428993  2017-06-21 00:06:44 +0000                13   
81   877201837425926144  2017-06-20 16:29:50 +0000                12   
82   876838120628539392  2017-06-19 16:24:33 +0000                12   
83   876537666061221889  2017-06-18 20:30:39 +0000                14   
84   876484053909872640  2017-06-18 16:57:37 +0000                13   
85   876120275196170240  2017-06-17 16:52:05 +0000                13   
86   875747767867523072  2017-06-16 16:11:53 +0000                13   
87   875144289856114688  2017-06-15 00:13:52 +0000                13   
88   875097192612077568  2017-06-14 21:06:43 +0000                13   
89   875021211251597312  2017-06-14 16:04:48 +0000                12   
90   874680097055178752  2017-06-13 17:29:20 +0000                12   
92   874296783580663808  2017-06-12 16:06:11 +0000                13   
93   874057562936811520  2017-06-12 00:15:36 +0000                12   
94   874012996292530176  2017-06-11 21:18:31 +0000                13   
96   873580283840344065  2017-06-10 16:39:04 +0000                13   
98   873213775632977920  2017-06-09 16:22:42 +0000                12   
99   872967104147763200  2017-06-09 00:02:31 +0000                12   
100  872820683541237760  2017-06-08 14:20:41 +0000                13   
102  872620804844003328  2017-06-08 01:06:27 +0000                13   
103  872486979161796608  2017-06-07 16:14:40 +0000                12   
104  872261713294495745  2017-06-07 01:19:32 +0000     

In [19]:
# single out first 19 digits of the timestamp
archive_df_clean['timestamp'] = archive_df_clean.timestamp.str[:19]

archive_df_clean['timestamp'] = pd.to_datetime(archive_df_clean['timestamp'], format = "%Y-%m-%d %H:%M:%S")

In [21]:
archive_df_clean.head(100)

tweet_id           timestamp  rating_numerator  \
0    892420643555336193 2017-08-01 16:23:56                13   
1    892177421306343426 2017-08-01 00:17:27                13   
2    891815181378084864 2017-07-31 00:18:03                12   
3    891689557279858688 2017-07-30 15:58:51                13   
4    891327558926688256 2017-07-29 16:00:24                12   
5    891087950875897856 2017-07-29 00:08:17                13   
6    890971913173991426 2017-07-28 16:27:12                13   
7    890729181411237888 2017-07-28 00:22:40                13   
8    890609185150312448 2017-07-27 16:25:51                13   
9    890240255349198849 2017-07-26 15:59:51                14   
10   890006608113172480 2017-07-26 00:31:25                13   
11   889880896479866881 2017-07-25 16:11:53                13   
12   889665388333682689 2017-07-25 01:55:32                13   
13   889638837579907072 2017-07-25 00:10:02                12   
14   889531135344209921 2017-07-24 17:02:04                13   
15   889278841981685760 2017-07-24 00:19:32                13   
16   888917238123831296 2017-07-23 00:22:39                12   
17   888804989199671297 2017-07-22 16:56:37                13   
18   888554962724278272 2017-07-22 00:23:06                13   
20   888078434458587136 2017-07-20 16:49:33                12   
21   887705289381826560 2017-07-19 16:06:48                13   
22   887517139158093824 2017-07-19 03:39:09                14   
23   887473957103951883 2017-07-19 00:47:34                13   
24   887343217045368832 2017-07-18 16:08:03                13   
25   887101392804085760 2017-07-18 00:07:08                12   
26   886983233522544640 2017-07-17 16:17:36                13   
27   886736880519319552 2017-07-16 23:58:41                13   
28   886680336477933568 2017-07-16 20:14:00                13   
29   886366144734445568 2017-07-15 23:25:31                12   
30   886267009285017600 2017-07-15 16:51:35                12   
..                  ...                 ...               ...   
76   878057613040115712 2017-06-23 01:10:23                14   
77   877736472329191424 2017-06-22 03:54:17                13   
79   877556246731214848 2017-06-21 15:58:08                12   
80   877316821321428993 2017-06-21 00:06:44                13   
81   877201837425926144 2017-06-20 16:29:50                12   
82   876838120628539392 2017-06-19 16:24:33                12   
83   876537666061221889 2017-06-18 20:30:39                14   
84   876484053909872640 2017-06-18 16:57:37                13   
85   876120275196170240 2017-06-17 16:52:05                13   
86   875747767867523072 2017-06-16 16:11:53                13   
87   875144289856114688 2017-06-15 00:13:52                13   
88   875097192612077568 2017-06-14 21:06:43                13   
89   875021211251597312 2017-06-14 16:04:48                12   
90   874680097055178752 2017-06-13 17:29:20                12   
92   874296783580663808 2017-06-12 16:06:11                13   
93   874057562936811520 2017-06-12 00:15:36                12   
94   874012996292530176 2017-06-11 21:18:31                13   
96   873580283840344065 2017-06-10 16:39:04                13   
98   873213775632977920 2017-06-09 16:22:42                12   
99   872967104147763200 2017-06-09 00:02:31                12   
100  872820683541237760 2017-06-08 14:20:41                13   
102  872620804844003328 2017-06-08 01:06:27                13   
103  872486979161796608 2017-06-07 16:14:40                12   
104  872261713294495745 2017-06-07 01:19:32                13   
105  872122724285648897 2017-06-06 16:07:15                12   
106  871879754684805121 2017-06-06 00:01:46                13   
107  871762521631449091 2017-06-05 16:15:56                12   
108  871515927908634625 2017-06-04 23:56:03                12   
110  871102520638267392 2017-06-03 20:33:19                14   
111  871032628920680449 2017-06-03 15:55:36       

#### Test

In [22]:
archive_df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 9 columns):
tweet_id              2175 non-null object
timestamp             2175 non-null datetime64[ns]
rating_numerator      2175 non-null int64
rating_denominator    2175 non-null int64
name                  2175 non-null object
doggo                 2175 non-null object
floofer               2175 non-null object
pupper                2175 non-null object
puppo                 2175 non-null object
dtypes: datetime64[ns](1), int64(2), object(6)
memory usage: 169.9+ KB


Issue #5: Non-Descriptive column header in image dataframe
#### Define:
change following column headers:

- p1 -> prediction_1
- p1_conf -> prediction_1_confidence
- p1_dog -> prediction_1_result
- p2 -> prediction_2
- p2_conf -> prediction_2_confidence
- p2_dog -> prediction_2_result
- p3 -> prediction_3
- p3_conf -> prediction_3_confidence
- p3_dog -> prediction_3_result
#### Code

In [23]:
image_df_clean = image_df.copy()
image_df_clean.rename(columns={'p1': 'prediction_1', 'p1_conf': 'prediction_1_confidence', 'p1_dog': 'prediction_1_result'}, inplace=True)
image_df_clean.rename(columns={'p2': 'prediction_2', 'p2_conf': 'prediction_2_confidence', 'p2_dog': 'prediction_2_result'}, inplace=True)
image_df_clean.rename(columns={'p3': 'prediction_3', 'p3_conf': 'prediction_3_confidence', 'p3_dog': 'prediction_3_result'}, inplace=True)

#### Test

In [24]:
image_df_clean.sample(100)

tweet_id                                            jpg_url  \
525   676606785097199616    https://pbs.twimg.com/media/CWPKSGpWcAQN6mw.jpg   
424   674063288070742018    https://pbs.twimg.com/media/CVrA-rIWEAANxwQ.jpg   
622   680497766108381184    https://pbs.twimg.com/media/CXGdG0aWcAEbOO1.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
1477  780543529827336192    https://pbs.twimg.com/media/CtUMLzRXgAAbZK5.jpg   
1296  752173152931807232    https://pbs.twimg.com/media/CnBBfNuWcAAkOgO.jpg   
735   687096057537363968    https://pbs.twimg.com/media/CYkON6CVAAAPXAc.jpg   
696   684481074559381504    https://pbs.twimg.com/media/CX_D6AJWwAAnBIw.jpg   
414   674008982932058114    https://pbs.twimg.com/media/CVqPkVoU4AAkXA7.jpg   
2005  877556246731214848    https://pbs.twimg.com/media/DC20wEcW0AAf59m.jpg   
509   676101918813499392    https://pbs.twimg.com/media/CWH_FTgWIAAwOUy.jpg   
2067  890729181411237888    https://pbs.twimg.com/media/DFyBahAVwAAhUTd.jpg   
160   668902994700836864    https://pbs.twimg.com/media/CUhruUgUAAAa8FQ.jpg   
1278  750101899009982464    https://pbs.twimg.com/media/Cmjlsh1XgAEvhq_.jpg   
308   671536543010570240    https://pbs.twimg.com/media/CVHG6_AWwAEJf_u.jpg   
715   685547936038666240    https://pbs.twimg.com/media/CYOONfZW8AA7IOA.jpg   
315   671735591348891648    https://pbs.twimg.com/media/CVJ79MzW4AEpTom.jpg   
135   668496999348633600    https://pbs.twimg.com/media/CUb6ebKWcAAJkd0.jpg   
869   697616773278015490    https://pbs.twimg.com/media/Ca5uv7RVAAA_QEg.jpg   
1449  776201521193218049    https://pbs.twimg.com/media/CsWfKadWEAAtmlS.jpg   
973   706681918348251136    https://pbs.twimg.com/media/Cc6jcYRXIAAFuox.jpg   
1157  733482008106668032    https://pbs.twimg.com/media/Ci3Z_idUkAA8RUh.jpg   
2032  883360690899218434    https://pbs.twimg.com/media/DEJT3FeXoAAtwUy.jpg   
1346  759197388317847553    https://pbs.twimg.com/media/Cok1_sjXgAU3xpp.jpg   
1554  793180763617361921    https://pbs.twimg.com/media/CwHxsdYVMAAqGCJ.jpg   
1737  821886076407029760    https://pbs.twimg.com/media/C2ftAxnWIAEUdAR.jpg   
344   672267570918129665    https://pbs.twimg.com/media/CVRfyZxWUAAFIQR.jpg   
800   691416866452082688    https://pbs.twimg.com/media/CZhn-QAWwAASQan.jpg   
2034  883838122936631299    https://pbs.twimg.com/media/DEQGFgAXUAAEvfi.jpg   
434   674372068062928900    https://pbs.twimg.com/media/CVvZ0KTWwAAdXKV.jpg   
...                  ...                                                ...   
1442  775364825476165632    https://pbs.twimg.com/media/CsKmMB2WAAAXcAy.jpg   
208   669970042633789440    https://pbs.twimg.com/media/CUw2MV4XIAAHLO_.jpg   
931   703079050210877440    https://pbs.twimg.com/media/CcHWqQCW8AEb0ZH.jpg   
793   690735892932222976    https://pbs.twimg.com/media/CZX8nyeVAAEstKM.jpg   
1705  817423860136083457  https://pbs.twimg.com/ext_tw_video_thumb/81742...   
361   672828477930868736    https://pbs.twimg.com/media/CVZd7ttWcAEs2wP.jpg   
1326  757354760399941633    https://pbs.twimg.com/media/CoKqIndWgAAattd.jpg   
998   708469915515297792    https://pbs.twimg.com/media/CdT9n7mW0AQcpZU.jpg   
1102  721001180231503872    https://pbs.twimg.com/media/CgGCvxAUkAAx55r.jpg   
1282  750383411068534784    https://pbs.twimg.com/media/CmnluwbXEAAqnkw.jpg   
1813  833479644947025920    https://pbs.twimg.com/media/C5EdT4jWEAARv2C.jpg   
1002  708810915978854401    https://pbs.twimg.com/media/CdYzwuYUIAAHPkB.jpg   
1814  833722901757046785    https://pbs.twimg.com/media/C5H6jmgW8AAevqq.jpg   
918   701601587219795968    https://pbs.twimg.com/media/CbyW7B0W8AIX8kX.jpg   
1933  859607811541651456    https://pbs.twimg.com/media/C-3wvtxXcAUTuBE.jpg   
1012  709519240576036864    https://pbs.twimg.com/media/Cdi3-f7W8AUOm9T.jpg   
658   682389078323662849    https://pbs.twimg.com/media/CXhVKtvW8AAyiyK.jpg   
2010  878776093423087618    https://pbs.twimg.com/media/DDIKMXzW0AEibje.jpg   
1018  710117014656950272    https://pbs.twimg.com/media/CdrXp9dWoA

Issue #6: the first letter of all dog names are not all capitalized.
#### Define:
capitalize the first letter of all dog breeds

#### Code

In [25]:
image_df_clean.prediction_1 = image_df_clean.prediction_1.str.title()
image_df_clean.prediction_2 = image_df_clean.prediction_2.str.title()
image_df_clean.prediction_3 = image_df_clean.prediction_3.str.title()


#### Test

In [26]:
image_df_clean.sample(100)

tweet_id                                            jpg_url  \
1289  751251247299190784  https://pbs.twimg.com/ext_tw_video_thumb/75125...   
754   688211956440801280  https://pbs.twimg.com/ext_tw_video_thumb/68821...   
1001  708738143638450176    https://pbs.twimg.com/media/CdXxlFPWwAABaOv.jpg   
884   698953797952008193    https://pbs.twimg.com/media/CbMuxV5WEAAIBjy.jpg   
520   676533798876651520    https://pbs.twimg.com/media/CWOH4s9U8AEtkmQ.jpg   
1432  773308824254029826    https://pbs.twimg.com/media/CrtYRMEWIAAUkCl.jpg   
1797  831552930092285952    https://pbs.twimg.com/media/C4pE-I0WQAABveu.jpg   
976   707021089608753152    https://pbs.twimg.com/media/Cc_XtkRW8AEE7Fn.jpg   
733   686749460672679938    https://pbs.twimg.com/media/CYfS75fWAAAllde.jpg   
145   668625577880875008    https://pbs.twimg.com/media/CUdvambWoAA007z.jpg   
1898  850753642995093505    https://pbs.twimg.com/media/C8576jrW0AEYWFy.jpg   
1488  782305867769217024    https://pbs.twimg.com/media/CttPBt0WIAAcsDE.jpg   
606   679877062409191424    https://pbs.twimg.com/media/CW9olDsUsAA0XSf.jpg   
1678  813202720496779264    https://pbs.twimg.com/media/C0kTjqIXgAAqpRi.jpg   
552   677565715327688705    https://pbs.twimg.com/media/CWcybBmWcAAigAQ.jpg   
1920  856282028240666624    https://pbs.twimg.com/media/C-If9ZwXoAAfDX2.jpg   
2010  878776093423087618    https://pbs.twimg.com/media/DDIKMXzW0AEibje.jpg   
1505  785264754247995392    https://pbs.twimg.com/media/CuXSHNnWcAIWEwn.jpg   
1167  735648611367784448    https://pbs.twimg.com/media/CjWMezdW0AErwU3.jpg   
1674  813142292504645637    https://pbs.twimg.com/media/C0jcmOKVQAAd0VR.jpg   
1043  712668654853337088    https://pbs.twimg.com/media/CePoVTyWsAQEz1g.jpg   
54    667044094246576128    https://pbs.twimg.com/media/CUHREBXXAAE6A9b.jpg   
1105  722974582966214656    https://pbs.twimg.com/media/CgiFjIpWgAA4wVp.jpg   
1172  736736130620620800    https://pbs.twimg.com/media/CjlpmZaUgAED54W.jpg   
1839  837482249356513284    https://pbs.twimg.com/media/C59VqMUXEAAzldG.jpg   
1675  813157409116065792    https://pbs.twimg.com/media/C0jqVVOXUAAGJ0G.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
1368  761750502866649088    https://pbs.twimg.com/media/CYLDikFWEAAIy1y.jpg   
396   673656262056419329    https://pbs.twimg.com/media/CVlOy3pW4AQ9H1K.jpg   
1428  772826264096874500    https://pbs.twimg.com/media/CrmhYYIXEAEcyYY.jpg   
...                  ...                                                ...   
1733  821407182352777218  https://pbs.twimg.com/ext_tw_video_thumb/82140...   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
548   677334615166730240    https://pbs.twimg.com/media/CWZgPPUWUAAUOvu.jpg   
728   686386521809772549    https://pbs.twimg.com/media/CYaI5aaW8AE8Uyk.jpg   
729   686606069955735556    https://pbs.twimg.com/media/CYdQktMWsAEI29_.jpg   
911   700796979434098688    https://pbs.twimg.com/media/Cbm7IeUXIAA6Lc-.jpg   
61    667152164079423490    https://pbs.twimg.com/media/CUIzWk_UwAAfUNq.jpg   
847   695409464418041856    https://pbs.twimg.com/media/CaaXN5LUYAEzAh-.jpg   
1695  816336735214911488    https://pbs.twimg.com/media/C1Q17WdWEAAjKFO.jpg   
841   694905863685980160    https://pbs.twimg.com/media/CaTNMUgUYAAB6vs.jpg   
578   678755239630127104    https://pbs.twimg.com/media/CWtsSQAUkAAnWws.jpg   
1569  794355576146903043    https://pbs.twimg.com/media/CvJCabcWgAIoUxW.jpg   
1327  757393109802180609    https://pbs.twimg.com/media/CoLNAq6WAAAkmdJ.jpg   
905   700151421916807169    https://pbs.twimg.com/media/CbdwATgWwAABGID.jpg   
422   674051556661161984    https://pbs.twimg.com/media/CVq2UHwWEAAduMw.jpg   
249   670679630144274432    https://pbs.twimg.com/media/CU67jGSUkAAk_1Y.jpg   
654   682059653698686977    https://pbs.twimg.com/media/CXcpovWWMAAMcfv.jpg   
2041  885311592912609280    https://pbs.twimg.com/media/C4bTH6nWMAAX_bJ.jpg   
939   703769065844768768    https://pbs.twimg.com/media/CcRKOzyXEA

Issue #7: Many ratings' denominator are not 10
#### Define:
Scale tweets who's denominators more than 10 or less than 10 to 10

#### Code

In [27]:
archive_df_clean.info()
archive_df_clean.rating_denominator.value_counts()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 9 columns):
tweet_id              2175 non-null object
timestamp             2175 non-null datetime64[ns]
rating_numerator      2175 non-null int64
rating_denominator    2175 non-null int64
name                  2175 non-null object
doggo                 2175 non-null object
floofer               2175 non-null object
pupper                2175 non-null object
puppo                 2175 non-null object
dtypes: datetime64[ns](1), int64(2), object(6)
memory usage: 169.9+ KB


10     2153
50        3
80        2
11        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

In [36]:
for index, row in archive_df_clean.iterrows():
    if row['rating_denominator'] == 0:
        new_denominator = 10
        archive_df_clean.set_value(index, 'rating_denominator', new_denominator)
    elif row['rating_denominator'] != 10:
        print(row['tweet_id'])
        new_denominator = 10
        new_numerator = round((row['rating_numerator']/row['rating_denominator'])*10, 1)
        print(row['rating_denominator'])
        print(new_denominator)
        print(row['rating_numerator'])
        print(new_numerator)
        archive_df_clean.set_value(index, 'rating_denominator', new_denominator)
        archive_df_clean.set_value(index, 'rating_numerator', new_numerator)
#     else:
#         new_denominator = rating_denominator
#         new_numerator = rating_numerator
    

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ipykernel_launcher.py:4: FutureWarning: set_value is deprecated and will be removed in a future release. Please use .at[] or .iat[] accessors instead
  after removing the cwd from sys.path.


#### Test

In [37]:
archive_df_clean.rating_denominator.value_counts()

10    2175
Name: rating_denominator, dtype: int64

Issue #8: Missing data
#### Define:
inner join 3 dataframe and make sure a tweet record has information in all 3 dataframe. 

#### Code#

In [ ]:
archive_df_clean.info()
image_df_clean.info()
status_df.info()
archive_df_clean_1 = archive_df_clean.copy()
type(archive_df_clean)


In [ ]:
# status_df_clean = status_df.copy()

# short_ID_List = image_df_clean['tweet_id']
# short_ID_List = image_df_clean.astype(str)

# archive_df_clean = archive_df_clean[archive_df_clean['tweet_id'].isin(short_ID_List)]
# status_df_clean = status_df_clean[status_df_clean["tweet_id"].isin(short_ID_List)]






In [ ]:


archive_df_clean.info()
status_df_clean.info()
#type(short_ID_List['tweet_id'][1])
#type(status_df_clean["tweet_id"][1])
#type(archive_df_clean['tweet_id'][1])

tidyness issue #2: dog stages not in on column
#### Define
combine doggo, floofer, pupper and puppo columns into stage column

#### Code



In [ ]:
#archive_df_clean['stage'] = archive_df_clean['doggo'] + archive_df_clean['floofer'] +  archive_df_clean['pupper'] + archive_df_clean['puppo'] 

for index, row in archive_df_clean.iterrows():
    if row["doggo"]  == "doggo" or row["floofer"] == "floofer" or row["pupper"] == "pupper" or row["puppo"] == ["puppo"]:
        archive_df_clean
        

#### Test

In [38]:
archive_df_clean.sample(100)

tweet_id           timestamp  rating_numerator  \
724   782747134529531904 2016-10-03 01:00:34                11   
62    880095782870896641 2017-06-28 16:09:20                11   
83    876537666061221889 2017-06-18 20:30:39                14   
2210  668620235289837568 2015-11-23 02:41:01                10   
1731  679844490799091713 2015-12-24 02:02:12                10   
45    883482846933004288 2017-07-08 00:28:19                 5   
1133  728751179681943552 2016-05-07 00:59:55                10   
1322  706291001778950144 2016-03-06 01:31:11                11   
1854  675534494439489536 2015-12-12 04:35:48                11   
1464  694356675654983680 2016-02-02 03:08:26                10   
674   789599242079838210 2016-10-21 22:48:24                12   
141   864873206498414592 2017-05-17 16:00:15                14   
1210  715680795826982913 2016-03-31 23:22:53                12   
916   756998049151549440 2016-07-23 23:42:53                11   
703   785927819176054784 2016-10-11 19:39:28                12   
722   783085703974514689 2016-10-03 23:25:55                11   
1701  680940246314430465 2015-12-27 02:36:20                 4   
2027  671874878652489728 2015-12-02 02:13:48                10   
1034  745057283344719872 2016-06-21 00:54:33                12   
1099  735991953473572864 2016-05-27 00:32:10                11   
1416  698635131305795584 2016-02-13 22:29:29                12   
878   760656994973933572 2016-08-03 02:02:14                11   
226   848690551926992896 2017-04-03 00:16:10                12   
1318  706538006853918722 2016-03-06 17:52:42                11   
912   757596066325864448 2016-07-25 15:19:12                 4   
1877  675109292475830276 2015-12-11 00:26:12                 9   
1870  675149409102012420 2015-12-11 03:05:37                12   
990   748705597323898880 2016-07-01 02:31:39                13   
1272  709225125749587968 2016-03-14 03:50:21                11   
1257  710296729921429505 2016-03-17 02:48:31                12   
...                  ...                 ...               ...   
106   871879754684805121 2017-06-06 00:01:46                13   
482   815390420867969024 2017-01-01 02:53:20                13   
30    886267009285017600 2017-07-15 16:51:35                12   
2346  666058600524156928 2015-11-16 01:01:59                 8   
1993  672609152938721280 2015-12-04 02:51:33                 9   
1779  677716515794329600 2015-12-18 05:06:23                12   
13    889638837579907072 2017-07-25 00:10:02                12   
1225  714141408463036416 2016-03-27 17:25:54                10   
848   766069199026450432 2016-08-18 00:28:24                12   
1157  724405726123311104 2016-04-25 01:12:38                10   
838   767500508068192258 2016-08-21 23:15:55                12   
348   831670449226514432 2017-02-15 01:04:21                13   
110   871102520638267392 2017-06-03 20:33:19                14   
1952  673680198160809984 2015-12-07 01:47:30                13   
262   842765311967449089 2017-03-17 15:51:22                12   
1352  703774238772166656 2016-02-28 02:50:28                10   
806   772102971039580160 2016-09-03 16:04:27                11   
1984  672902681409806336 2015-12-04 22:17:55                 9   
1138  728035342121635841 2016-05-05 01:35:26                12   
2021  672082170312290304 2015-12-02 15:57:30                11   
1801  676957860086095872 2015-12-16 02:51:45                10   
1190  718234618122661888 2016-04-08 00:30:51                11   
364   829141528400556032 2017-02-08 01:35:19                12   
1452  695767669421768709 2016-02-06 00:35:13                13   
508   812503143955202048 2016-12-24 03:40:19                13   
1624  684880619965411328 2016-01-06 23:33:58                 5   
445   819227688460238848 2017-01-11 17:01:16                12   
2105  670474236058800128 2015-11-28 05:28:09                10   
1996  672591762242805761 2015-12-04 01:42:26            

# Analysis and Visualization

In [ ]:
import matplotlib as mp



# reference

1. multiple json in txt line by line: https://stackoverflow.com/questions/17055117/python-json-dump-append-to-txt-with-each-variable-on-new-line
2. load json from txt line by line: https://stackoverflow.com/questions/21533894/how-to-read-line-delimited-json-from-large-file-line-by-line
3. create dataframe from json file: https://stackoverflow.com/questions/47925828/how-to-create-a-pandas-dataframe-using-tweepy
4. keep only rows who has NaN in a certain column: https://stackoverflow.com/questions/25430995/keeping-nan-values-and-dropping-nonmissing-values
5. str to datetime: https://stackoverflow.com/questions/17134716/convert-dataframe-column-type-from-string-to-datetime
6. iterate rows to change value https://stackoverflow.com/questions/23330654/update-a-dataframe-in-pandas-while-iterating-row-by-row
7. use isin
8. combine to different columns into 1: 